### PR TITLE
[feat][ws] PIP-290 Make WSS support E2E encryption

### PR DESCRIPTION
**Pull request by poorbarcode**
### Motivation & Modifications

See https://github.com/apache/pulsar/pull/20923

### Documentation

- [ ] `doc` 
- [x] `doc-required` 
- [ ] `doc-not-needed` 
- [ ] `doc-complete` 

### Matching PR in forked repository

PR in forked repository: x

### CODE REVIEW

File: pulsar-websocket/src/main/java/org/apache/pulsar/websocket/data/ProducerMessage.java [hunk 1]
```diff
@@ -69,4 +70,13 @@ public class ProducerMessage {
...
     // Base64 encoded serialized schema for payload
     public String valueSchema;
+
+    // Compression type. Do not set it if compression is not performed.
+    public String encryptionParam;
+
+    // The size of the payload before compression. Do not set it if compression is not performed.
+    public CompressionType compressionType;
+
+    // Base64 encoded serialized initialization vector used when the client encrypts.
```
Comment (codelipenghui) [inline pulsar-websocket/src/main/java/org/apache/pulsar/websocket/data/ProducerMessage.java:80]: It's an integer, shouldn't be base64 encoded?

File: pulsar-websocket/src/main/java/org/apache/pulsar/websocket/data/ProducerMessage.java [hunk 2]
```diff
@@ -69,4 +70,13 @@ public class ProducerMessage {
 
     // Base64 encoded serialized schema for payload
     public String valueSchema;
+
+    // Compression type. Do not set it if compression is not performed.
```
Comment (codelipenghui) [inline pulsar-websocket/src/main/java/org/apache/pulsar/websocket/data/ProducerMessage.java:74]: It's not compression type

File: pulsar-websocket/src/main/java/org/apache/pulsar/websocket/data/ProducerMessage.java [hunk 3]
```diff
@@ -69,4 +70,13 @@ public class ProducerMessage {
...
+    public String encryptionParam;
+
+    // The size of the payload before compression. Do not set it if compression is not performed.
```
Comment (codelipenghui) [inline pulsar-websocket/src/main/java/org/apache/pulsar/websocket/data/ProducerMessage.java:77]: Ah, this one is for `uncompressedMessageSize`

File: pulsar-websocket/src/main/java/org/apache/pulsar/websocket/data/ConsumerMessage.java [hunk 4]
```diff
@@ -35,4 +35,5 @@ public class ConsumerMessage {
     public EncryptionContext encryptionContext;
 
     public String key;
+    // TODO support ackSet in the future.
```
Comment (codelipenghui) [inline pulsar-websocket/src/main/java/org/apache/pulsar/websocket/data/ConsumerMessage.java:38]: We can remove it first. Not sure if it is necessary to support ackSet.

File: pulsar-websocket/src/main/java/org/apache/pulsar/websocket/ProducerHandler.java [hunk 5]
```diff
@@ -315,30 +351,117 @@ protected ProducerBuilder<byte[]> getProducerBuilder(PulsarClient client) {
...
-        if (queryParams.containsKey("batchingEnabled")) {
-            builder.enableBatching(Boolean.parseBoolean(queryParams.get("batchingEnabled")));
+        if (queryParams.containsKey("messageRoutingMode")) {
+            checkArgument(
+                    Enums.getIfPresent(MessageRoutingMode.class, queryParams.get("messageRoutingMode")).isPresent(),
+                    "Invalid messageRoutingMode %s", queryParams.get("messageRoutingMode"));
+            MessageRoutingMode routingMode = MessageRoutingMode.valueOf(queryParams.get("messageRoutingMode"));
+            if (!MessageRoutingMode.CustomPartition.equals(routingMode)) {
+                builder.messageRoutingMode(routingMode);
+            }
```
Comment (codelipenghui) [inline pulsar-websocket/src/main/java/org/apache/pulsar/websocket/ProducerHandler.java:361]: Is it a part of the proposal?

File: pulsar-websocket/src/main/java/org/apache/pulsar/websocket/ProducerHandler.java [hunk 6]
```diff
@@ -315,30 +351,117 @@ protected ProducerBuilder<byte[]> getProducerBuilder(PulsarClient client) {
...
+                        .map(e -> new KeyValue().setKey(e.getKey()).setValue(e.getValue()))
+                        .collect(Collectors.toList());
+            }
+            builder.addEncryptionKey(keyNameArray[index]);
         }
+        // Disable server-side batch process, compression, and encryption, and only set the message metadata that
+        // which client-side provided.
+        builder.enableBatching(false);
+        builder.compressionType(CompressionType.NONE);
+        builder.cryptoKeyReader(DummyCryptoKeyReaderImpl.INSTANCE);
```
Comment (codelipenghui) [inline pulsar-websocket/src/main/java/org/apache/pulsar/websocket/ProducerHandler.java:429]: Please add a comment for these two lines to explain why we are using NONE compressor and DummyCryptoKeyReader.

File: pulsar-websocket/src/main/java/org/apache/pulsar/websocket/ProducerHandler.java [hunk 7]
```diff
@@ -315,30 +351,117 @@ protected ProducerBuilder<byte[]> getProducerBuilder(PulsarClient client) {
...
+                    builder.maxPendingMessages(Integer.parseInt(queryParams.get("maxPendingMessages")));
+                }
+
+                if (queryParams.containsKey("batchingMaxPublishDelay")) {
+                    builder.batchingMaxPublishDelay(Integer.parseInt(queryParams.get("batchingMaxPublishDelay")),
+                            TimeUnit.MILLISECONDS);
+                }
+            } else {
+                builder.enableBatching(false);
+                printWarnLogIfSettingBatchedParams();
```
Comment (codelipenghui) [inline pulsar-websocket/src/main/java/org/apache/pulsar/websocket/ProducerHandler.java:464]: Should be removed?

File: pulsar-websocket/src/main/java/org/apache/pulsar/websocket/ProducerHandler.java [hunk 8]
```diff
@@ -356,7 +479,27 @@ protected ProducerBuilder<byte[]> getProducerBuilder(PulsarClient client) {
...
+        if (queryParams.containsKey("batchingMaxMessages")) {
+            log.warn("Since batchingEnabled is false, the param batchingMaxMessages of producer will be ignored");
+        }
+        if (queryParams.containsKey("maxPendingMessages")) {
+            log.warn("Since batchingEnabled is false, the param maxPendingMessages of producer will be ignored");
+        }
+        if (queryParams.containsKey("batchingMaxPublishDelay")) {
+            log.warn("Since batchingEnabled is false, the param batchingMaxPublishDelay of producer will be ignored");
+        }
+    }
```
Comment (codelipenghui) [inline pulsar-websocket/src/main/java/org/apache/pulsar/websocket/ProducerHandler.java:497]: It should be removed, right? It's not a part of the proposal

**Review by codelipenghui — CHANGES_REQUESTED** [review]

File: pulsar-websocket/src/main/java/org/apache/pulsar/websocket/ProducerHandler.java (hunk 7 above)
Comment (poorbarcode) [reply pulsar-websocket/src/main/java/org/apache/pulsar/websocket/ProducerHandler.java:464]: No, it is the original design: Web socket proxy will composite multi messages into a batched message if users set the param "batchingEnabled" as `true`. I just moved this code into the method `popularProducerBuilderForServerSideEncrypt()`

File: pulsar-websocket/src/main/java/org/apache/pulsar/websocket/ProducerHandler.java (hunk 6 above)
Comment (poorbarcode) [reply pulsar-websocket/src/main/java/org/apache/pulsar/websocket/ProducerHandler.java:429]: Done

File: pulsar-websocket/src/main/java/org/apache/pulsar/websocket/ProducerHandler.java (hunk 5 above)
Comment (poorbarcode) [reply pulsar-websocket/src/main/java/org/apache/pulsar/websocket/ProducerHandler.java:361]: Ah, it is the [original design](https://github.com/apache/pulsar/pull/862), I've not changed it.

See: https://github.com/apache/pulsar/pull/862

File: pulsar-websocket/src/main/java/org/apache/pulsar/websocket/data/ConsumerMessage.java (hunk 4 above)
Comment (poorbarcode) [reply pulsar-websocket/src/main/java/org/apache/pulsar/websocket/data/ConsumerMessage.java:38]: it was removed

File: pulsar-websocket/src/main/java/org/apache/pulsar/websocket/data/ProducerMessage.java (hunk 3 above)
Comment (poorbarcode) [reply pulsar-websocket/src/main/java/org/apache/pulsar/websocket/data/ProducerMessage.java:77]: Fixed

File: pulsar-websocket/src/main/java/org/apache/pulsar/websocket/data/ProducerMessage.java (hunk 1 above)
Comment (poorbarcode) [reply pulsar-websocket/src/main/java/org/apache/pulsar/websocket/data/ProducerMessage.java:80]: Fixed

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerImpl.java [hunk 9]
```diff
@@ -1739,7 +1739,7 @@ private ByteBuf decryptPayloadIfNeeded(MessageIdData messageId, int redeliveryCo
         if (conf.getCryptoKeyReader() == null) {
             switch (conf.getCryptoFailureAction()) {
                 case CONSUME:
-                    log.warn("[{}][{}][{}] CryptoKeyReader interface is not implemented. Consuming encrypted message.",
+                    log.info("[{}][{}][{}] CryptoKeyReader interface is not implemented. Consuming encrypted message.",
```
[the remark this answers is not in the document]
Comment (poorbarcode) [reply pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerImpl.java:1742]: Done

File: pulsar-broker/src/test/java/org/apache/pulsar/websocket/proxy/WssClientSideEncryptUtils.java [hunk 10]
```diff
@@ -0,0 +1,342 @@
...
+
+    private static final FastThreadLocal<SingleMessageMetadata> LOCAL_SINGLE_MESSAGE_METADATA =
+            new FastThreadLocal<>() {
+                @Override
+                protected SingleMessageMetadata initialValue() throws Exception {
+                    return new SingleMessageMetadata();
+                }
+            };
+
+    public static byte[] compositeBatchMessage(List<ProducerMessage> messages) {
```
[the remark this answers is not in the document]
Comment (poorbarcode) [reply pulsar-broker/src/test/java/org/apache/pulsar/websocket/proxy/WssClientSideEncryptUtils.java:216]: It was removed

File: pulsar-websocket/src/main/java/org/apache/pulsar/websocket/ProducerHandler.java (hunk 8 above)
Comment (poorbarcode) [reply pulsar-websocket/src/main/java/org/apache/pulsar/websocket/ProducerHandler.java]: The parameters `batchingMaxMessages`, `batchingMaxMessages`, `maxPendingMessages`, `batchingMaxPublishDelay`, and `compressionType` will be discarded when a WSS producer is registering with a JSON typed `encryptionKeys`. Print these logs to let users know these parameters will not be used. Maybe print a INFO level log is better?

And this behavior was explained in the PIP https://github.com/apache/pulsar/pull/20923/files#diff-bdaca0e5fd3a5a0847fa77db82b66797d718cf469c90e81a0755a7e8eaa67c2fR96-R97

Comment (poorbarcode) [reply pulsar-websocket/src/main/java/org/apache/pulsar/websocket/ProducerHandler.java]: Changed the log level to `INFO`

File: pulsar-websocket/src/main/java/org/apache/pulsar/websocket/ProducerHandler.java (hunk 5 above)
Comment (codelipenghui) [reply pulsar-websocket/src/main/java/org/apache/pulsar/websocket/ProducerHandler.java:361]: Got it.

**Review by codelipenghui — APPROVED** [review]

**Comment by codelipenghui** [comment]
It looks like the [failed test](https://github.com/apache/pulsar/actions/runs/5935884541/job/16095506330?pr=20958) is related to the change?

File: pulsar-websocket/src/main/java/org/apache/pulsar/websocket/ProducerHandler.java [hunk 11]
```diff
@@ -192,6 +207,27 @@ public void onWebSocketText(String message) {
...
+                }
+            } catch (Exception e){
+                String msg = format("Invalid Base64 encryptionParam error=%s", e.getMessage());
+                sendAckResponse(new ProducerAck(PayloadEncodingError, msg, null, requestContext));
+                return;
+            }
+            if (sendRequest.compressionType != null && sendRequest.uncompressedMessageSize != null) {
+                // Set compression information.
+                builder.getMetadataBuilder().setCompression(sendRequest.compressionType);
+                builder.getMetadataBuilder().setUncompressedSize(sendRequest.uncompressedMessageSize);
```
Comment (mattisonchao) [inline pulsar-websocket/src/main/java/org/apache/pulsar/websocket/ProducerHandler.java:227]: Should we clarify this behaviour?  or add some checks?
- If the user sets `compressionType` without `unCompressedSize` the data will auto compressed by the broker.
- How about the user only set uncompressed size but not set compression type?

File: pulsar-websocket/src/main/java/org/apache/pulsar/websocket/ProducerHandler.java [hunk 12]
```diff
@@ -206,7 +242,7 @@ public void onWebSocketText(String message) {
             }
         }).exceptionally(exception -> {
             log.warn("[{}] Error occurred while producer handler was sending msg from {}: {}", producer.getTopic(),
-                    getRemote().getInetSocketAddress().toString(), exception.getMessage());
+                    getRemote().getInetSocketAddress().toString(), exception.getMessage(), exception);
```
Comment (mattisonchao) [inline pulsar-websocket/src/main/java/org/apache/pulsar/websocket/ProducerHandler.java:245]: If we wanna print the exception stack trace, it's better remove the duplicated exception message?

File: pulsar-client-api/src/main/java/org/apache/pulsar/client/api/DummyCryptoKeyReaderImpl.java [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pulsar.client.api;
+
+import java.util.Map;
+
+public class DummyCryptoKeyReaderImpl implements CryptoKeyReader {
```
Comment (mattisonchao) [inline pulsar-client-api/src/main/java/org/apache/pulsar/client/api/DummyCryptoKeyReaderImpl.java:23]: It's better to add comments to explain why we need it.

**Review by mattisonchao — APPROVED** [review]
Overall LGTM, left some comments.

**Comment by mattisonchao** [comment]
Is it possible to add an integration test for this in the future? :)

File: pulsar-client-api/src/main/java/org/apache/pulsar/client/api/DummyCryptoKeyReaderImpl.java (hunk 13 above)
Comment (poorbarcode) [reply pulsar-client-api/src/main/java/org/apache/pulsar/client/api/DummyCryptoKeyReaderImpl.java:26]: Done

File: pulsar-websocket/src/main/java/org/apache/pulsar/websocket/ProducerHandler.java (hunk 12 above)
Comment (poorbarcode) [reply pulsar-websocket/src/main/java/org/apache/pulsar/websocket/ProducerHandler.java]: fixed

File: pulsar-websocket/src/main/java/org/apache/pulsar/websocket/ProducerHandler.java (hunk 11 above)
Comment (poorbarcode) [reply pulsar-websocket/src/main/java/org/apache/pulsar/websocket/ProducerHandler.java:227]: Added a check

**Comment by poorbarcode** [comment]
> Is it possible to add an integration test for this in the future? :)

In the test, the Web Socket Proxy is a real server, so I think this is already an integration test?

**Pull request merged**